abd-clam/src/core/dataset/vec2d.rs

```diff
@@ -43,284 +43,236 @@ impl<T: Send + Sync + Copy, U: Number> VecDataset<T, U> {
     /// * `metric`: The metric for computing distances between instances.
     /// * `is_expensive`: Whether the metric is expensive to compute.
-    pub fn new(name: String, data: Vec<T>, metric: fn(T, T) -> U, is_expensive: bool) -> Self {
-        let indices = (0..data.len()).collect();
+    pub fn new(name: String, data: Vec<I>, metric: fn(&I, &I) -> U, is_expensive: bool) -> Self {
         Self {
             name,
             data,
             metric,
             is_expensive,
-            indices,
-            reordering: None,
+            permuted_indices: None,
         }
     }
 }
 
-impl<'a, V: Number, U: Number> VecDataset<&'a [V], U> {
-    /// Saves a numeric dataset to a specified path in binary format
-    ///
-    /// # Errors
-    /// Errors on file write issues or if a reordered index does not exist
-    pub fn save(&self, path: &Path) -> Result<(), String> {
-        let mut handle = File::create(path).map_err(|err| format!("{err}"))?;
+impl<I: Instance, U: Number, Idx> Index<Idx> for VecDataset<I, U>
+where
+    Idx: SliceIndex<[I], Output = I>,
+{
+    type Output = Idx::Output;
 
-        // Write header (Basic protection against reading bad data)
-        handle.write_all(b"VEC2D").map_err(|err| format!("{err}"))?;
+    fn index(&self, index: Idx) -> &Self::Output {
+        self.data.index(index)
+    }
+}
 
-        // Write dataset name
-        handle.write_all(self.name.as_bytes()).map_err(|err| format!("{err}"))?;
+impl<I: Instance, U: Number> Dataset<I, U> for VecDataset<I, U> {
+    fn type_name(&self) -> String {
+        format!("VecDataset<{}>", I::type_name())
+    }
 
-        // Write null terminator
-        handle.write(&[0]).map_err(|err| format!("{err}"))?;
+    fn name(&self) -> &str {
+        &self.name
+    }
 
-        // Write cardinality
-        let cardinality = self.data.len();
-        let cardinality_bytes = cardinality.to_le_bytes();
-        handle.write_all(&cardinality_bytes).map_err(|err| format!("{err}"))?;
+    fn cardinality(&self) -> usize {
+        self.data.len()
+    }
 
-        // Write dimensionality
-        let dimensionality = self.data[0].len();
-        let dimensionality_bytes = dimensionality.to_le_bytes();
-        handle
-            .write_all(&dimensionality_bytes)
-            .map_err(|err| format!("{err}"))?;
+    fn is_metric_expensive(&self) -> bool {
+        self.is_expensive
+    }
 
-        // Write the type name for V
-        handle
-            .write_all(V::type_name().as_bytes())
-            .map_err(|err| format!("{err}"))?;
+    fn metric(&self) -> fn(&I, &I) -> U {
+        self.metric
+    }
 
-        // Write null terminator
-        handle.write(&[0]).map_err(|err| format!("{err}"))?;
+    fn permute_instances(&mut self, permutation: &[usize]) -> Result<(), String> {
+        let n = permutation.len();
+
+        // The "source index" represents the index that we hope to swap to
+        let mut source_index: usize;
```

Comment:
This naming is counterintuitive to me

Reply:
Eh, I guess it's fine. It's a "source" index in the sense that it's the source of the element that will be put in position i on this iteration